include/deal.II/lac/solver_gmres.h

```diff
@@ -687,28 +736,58 @@ class SolverFGMRES : public SolverBase<VectorType>
   /**
    * Constructor.
    */
-  SolverFGMRES(SolverControl            &cn,
-               VectorMemory<VectorType> &mem,
-               const AdditionalData     &data = AdditionalData());
+  SolverMPGMRES(SolverControl            &cn,
+                VectorMemory<VectorType> &mem,
+                const AdditionalData     &data = AdditionalData());
 
   /**
    * Constructor. Use an object of type GrowingVectorMemory as a default to
    * allocate memory.
    */
-  SolverFGMRES(SolverControl        &cn,
-               const AdditionalData &data = AdditionalData());
+  SolverMPGMRES(SolverControl        &cn,
+                const AdditionalData &data = AdditionalData());
 
   /**
    * Solve the linear system $Ax=b$ for x.
    */
-  template <typename MatrixType, typename PreconditionerType>
+  template <typename MatrixType, typename... PreconditionerTypes>
   DEAL_II_CXX20_REQUIRES(
     (concepts::is_linear_operator_on<MatrixType, VectorType> &&
-     concepts::is_linear_operator_on<PreconditionerType, VectorType>))
-  void solve(const MatrixType         &A,
-             VectorType               &x,
-             const VectorType         &b,
-             const PreconditionerType &preconditioner);
+     (concepts::is_linear_operator_on<PreconditionerTypes, VectorType> && ...)))
+  void solve(const MatrixType &A,
+             VectorType       &x,
+             const VectorType &b,
+             const PreconditionerTypes &...preconditioners);
+
+protected:
+  /**
+   * Indexing strategy to construct the search space.
+   *
+   * This enum class is internally used in the implementation of the
+   * MPGMRES algorithm to switch between the strategies.
+   */
+  enum class IndexingStrategy
+  {
+    fgmres,
+    truncated_mpgmres,
+    full_mpgmres,
+  };
+
+  /**
+   * The indexing strategy chosen for MPGMRES. This internal variable gets
+   * set by the constructor of SolverFGMRES and SolverMPGMRES.
+   */
+  IndexingStrategy indexing_strategy;
```

Comment:
As mentioned below, I think this would work better if you make it an argument of the `solve_internal` function, rather than part of the state of the entire object.

Reply:
I agree, that could make sense. But we will need to make the `AdditionalData` protected to allow it to be called by the derived class `FGMRES`.